Add ut to project and create tests for each of the algorithms.

Sebanisu opened the pull request:
Add https://github.com/boost-ext/ut to the project to test it, and lead to having a CI run on it like it does for J. This will require compiler upgrades on to at least (GCC-9+, Clang-9.0+, Apple Clang-11.0.0+, MSVC-2019+*, Clang-cl-9.0+) just for the tests to build. Though the algorithm.hpp should still include just fine, into older compilers, Until we add newer standard code to it.

- [x] add .gitignore (I guess we did have one already, I copied the one from the j project as it filters more things out)
- [x] ~~add `static_assert(sizeof...(Is) != 0U);` to transform and find. Since it's an error to not pass at least 1 iterator to these functions.~~ Turns out it's okay to only have 1 range if the op or pred is coded to handle one range.
- [x] set library to c++20 (The .hpp files should still work on c++17 till we add c++20 code) (required by ut)
- [x] add ut to cmake
- [x] add test executable
- [x] add tests for algorithm.hpp
  - [x] add some variadic operations. Such as:
    - [x] plus, multiply, to test transform,
    - [x] equal_to, not_equal_to for testing find, found
    - [x] make_signed, make_unsigned these take a op and values change them to all signed or all unsigned and then pass to the op and return the value. 
  - [x] make sure to have constexpr tests to make sure the code works at constexpr time.
    - [x] transform overload that takes a r-value for output range so I can pass a std::array and it returns the transformed version.
    - [x] iota, I know we are getting a views::iota but it's only in one compiler as of now.
- [x] add test executable to ctest with add_test
- [x] add compiler warnings to test executable.
- [x] import .clang-format, I copied the one from the j project
- [x] appveyor.yml - I added this as a CI option. Though you might wanna set up your own CI like you did with j. Appveyor has to update gcc, ninja and cmake every time it runs. :P And I just did linux we probably wanna have build scripts for windows and mac. I created this with the appveyor website and exported the script to a yml. https://ci.appveyor.com/project/Sebanisu/an-algorithm-library
  - [x] Build on MSVC 2019 - is the fastest because appveyor keeps MSVC up to date so we don't need to download updates.
  - [x] Build on Clang 12
  - [x] Build on GCC 10.2

I added extra functions to `algorithm_test.cpp` as I didn't want to add new functions to `algorithm.h` without checking with you first. These extra functions might be used in tests but might not be needed in the aal library.



resolves: https://github.com/codereport/An-Algorithm-Library/issues/3

Sebanisu committed:
copied the .gitignore from j project.

Sebanisu committed:
set library to c++20

This should only effect projects with cpp files.

Sebanisu committed:
create tests

Sebanisu committed:
Added compiler-warnings and rearranged files. added a couple tests

Sebanisu committed:
east const and added some tests reworded test names. changed custom transform to consteval

Sebanisu commented:
I think this is a good start. :)

Sebanisu committed:
instead of baking in a initializer list of numbers I wrote a consteval iota.

Though maybe iota::view would work better. it only works in gcc i think.

Sebanisu committed:
copied .clang-format to project from j. ran clang format on algorithm_test.cpp

Sebanisu committed:
Create CONTRIBUTING.md

based on the jsource one.

Sebanisu committed:
I guess there is supposed to be a NAME and COMMAND word on this .

Sebanisu commented:
~~`ninja -C build test` is working in my other project. And I based my cmake code on that and this one it isn't working. So something is off in my cmake.~~

fixed enable_testing() needs to be ran on the root CMakeLists.txt

Sebanisu committed:
enable_testing() needs to be in the root CMakeLists.txt or included via a .cmake file into the root.

Sebanisu committed:
removed note about command not working as it's working now.

Sebanisu committed:
remove command #

Sebanisu committed:
add command for verbose output from tests

Sebanisu committed:
add a custom executable name to algorithm_test

Sebanisu committed:
Create appveyor.yml

If you wanna try appveyor. I tried this build script out. It worked and ran tests. I'm not an expert though, so some of these commands might not be required.

Sebanisu committed:
you have to use the PRIVATE keyword for each library you include

herwinw commented on .gitignore:
```diff
@@ -1,4 +1,147 @@
+!asm/*.o
+
+# temp scripts used for debugging
+*makevs/jdll/*.ijs
```
A lot of these ignores are J-specific, which should not be relevant in this repository.

Sebanisu replied:
I think I got them all :) 

https://github.com/codereport/An-Algorithm-Library/pull/7/commits/6f8abcaa9541c282d88bf2111945087ac15539e3

herwinw commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,34 @@
+    mkdir build
+    cmake -G "Ninja Multi-Config" -B build
+    ```
+   * To build debug:
+    ```sh
+    ninja -C build 
+    ```
+   * To build release:
+    ```sh
+    ninja -C build -f build-Release.ninja
+    ```
````
Do we actually build something here? The only thing that should be compiled are the tests itself, which should be done in the test command.

Sebanisu replied:
Oh, I see what you mean I can take that out. Though if you use the verbose output it requires the test to be built first. So I just moved the build command infront of that.

https://github.com/codereport/An-Algorithm-Library/pull/7/commits/7a275d579f579332c6c797b00611e1a742a7aafd

Sebanisu committed:
fixed compiler warnings.

seems the warnings have to be added to a interface library to work correctly. I also had to block the shadow errors to compile ut.

Sebanisu committed:
Should probably be private, though for executables it probably doesn't matter.

Sebanisu committed:
herwinw pointed out we don't need to build anything.

Though I'm using the verbose output, and it needs to build.

Sebanisu committed:
remove j .gitignore entries.

Sebanisu committed:
wrap tests in if statement per "modern-cmake"

I found https://cliutils.gitlab.io/modern-cmake/chapters/testing.html

We need to put if statements around the enable testing and the add subdirectory.

Sebanisu commented:
I found https://cliutils.gitlab.io/modern-cmake/chapters/testing.html so I added a some if statements in https://github.com/codereport/An-Algorithm-Library/pull/7/commits/adb223c8dbad792108a7ec3f90f4844b4efcbd51

Without the if statements the cTest was still looking for the tests even if they weren't built. So j's tests would show failed because it would try to run An-Algorithm-Library's tests.

https://cmake.org/cmake/help/latest/variable/CMAKE_PROJECT_NAME.html
https://cmake.org/cmake/help/latest/variable/PROJECT_NAME.html

Sebanisu committed:
Setting target specific settings instead of global

Sebanisu committed:
appveyor to now build on visual studio 2019

Sebanisu committed:
add clang to CI

this is some trial and error. It seemed to work on my code except clang doesn't have <concepts> so it errored on that.

Sebanisu committed:
clang failed because it lacked <ranges> and <concepts>

we could probably swap out ranges code in places. though I am using a concept from ranges to check for a range.

Sebanisu committed:
fix clang? removed ranges.

though I can't get clang to fail on my computer so maybe it will on the CI

Sebanisu committed:
custom iota to fix clang?

Sebanisu committed:
just realized ninja command should be used to build msvc.

Sebanisu committed:
Revert "just realized ninja command should be used to build msvc."

This reverts commit dd5b38da2550832257242544098d44641ec7cf44.

Sebanisu committed:
add Windows instructions to the Contributing.md

Really though if you're on msvc you're probably opening the project with a gui. Though we have linux commandline so I added the windows one.

I found out on the CI if you use the same ninja commands as linux ninja will use clang for windows.

Sebanisu committed:
add how to swap between compilers on linux.

There might be a better way? I just put what I found seems to work.

Sebanisu committed:
formatted the md

Sebanisu committed:
reformated the end of this line.

Sebanisu committed:
created add_test_common function to cmake.

wrapped up the code used for making tests into a function so it's easier to add new tests.

Sebanisu committed:
added /wd4505

suppress unreferenced local function has been removed.

I had this issue where one test was using a function and another wasn't so visual studio wasn't compiling because I have all warnings are errors turned on. In gcc adding `[[maybe_unused]]` was enough to fix this kind of issue but that didn't help in msvc.

Sebanisu committed:
comment out warning suppression

probably won't need this warning suppressed. But the code is there if we end up needing it.

Sebanisu committed:
change found to any_of in the test.

Sebanisu commented:
Appveyor's clang support might be limited till I solve a build script issue. https://github.com/Sebanisu/ToolsLibrary/issues/6

Some c++20 features require libc++ to work like std::span. You might need to disable building on clang if you use too many new features of c++20.

Sebanisu commented:
MSVC has a bug with c++20 templated lambdas opened a ticket on microsoft https://github.com/Sebanisu/ToolsLibrary/issues/4 though it might not be an issue here.

Sebanisu committed:
clang 12 build script

Sebanisu committed:
move project_warnings library into the .cmake file and added clang compat cmake code.

Sebanisu commented:
I was able to get it to build on clang12 with libc++ Might not be the best build script but it's working :)

Sebanisu committed:
add GIT_REMOTE_UPDATE_STRATEGY

https://cmake.org/cmake/help/v3.18/module/ExternalProject.html because if you make local changes the default behavior is to stash them and pop the stash after pulling the update. So it'll cause an error if there is a merge conflict.
